### PR TITLE
Feature: text to speech on chat screens

### PR DESCRIPTION
**Pull request by Rolph1**
A simple PR to implement a feature which consists of an AI voice reading out the interviewer's responses when on the chat screen.

**Files changed**:
MainActivity.kt : Initialization of the text to speech class instance and setting the speed of the speech so as to not be too slow and boring.

ChatViewModel.kt : to transform the GPT response text to audio as soon as it is received in the chat screen, also to cut off the audio when exiting the screen

ChatViewModelTest.kt : added unit tests to make sure the feature of cutting off the audio of the AI works

### CODE REVIEW

**Comment by charbelraffoul** [comment]
Hi Rudolf,

I love this addition! It makes the interaction with the app much smoother and more engaging. However, I noticed that for each message, the voice starts with a sentence (I couldn’t catch exactly what it says), followed by "content =" before it begins the actual message. Could this be removed or adjusted? Also, I think the current speech pace is a bit fast—would it be possible to slow it down slightly for better clarity?

I agree with the decision not to add tests for this feature since it primarily complements the main functionality and doesn’t directly contribute to the UI or backend. Given the nature of the feature, adding tests would indeed be challenging and might not add significant value. Great call!

Great work overall! Maybe in the future, a good addition could be allowing the user to choose the voice they want to interact with. What do you think about this idea?

Looking forward to hearing your thoughts!

File: app/src/main/java/com/github/se/orator/MainActivity.kt [hunk 1]
```diff
@@ -97,6 +100,14 @@ class MainActivity : ComponentActivity() {
       }
     }
 
+    textToSpeech =
+        TextToSpeech(this) { status ->
+          if (status == TextToSpeech.SUCCESS) {
+            textToSpeech.setSpeechRate(1.7f)
```
Comment (charbelraffoul) [inline app/src/main/java/com/github/se/orator/MainActivity.kt:106]: Maybe we could adjust the speech rate to make it a bit lower. What do you think about this?

**Review by charbelraffoul — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/github/se/orator/MainActivity.kt (hunk 1 above)
Comment (Rolph1) [reply app/src/main/java/com/github/se/orator/MainActivity.kt:106]: Thanks for the recommendation Charbel! I have tried different speeds and came to the conclusion that this was the best. It's not so slow that you get bored and yet it's not so fast you can't keep up.
(Feel free to recommend a better value if you find one!!)

**Comment by Rolph1** [comment]
> Hi Rudolf,
> 
> I love this addition! It makes the interaction with the app much smoother and more engaging. However, I noticed that for each message, the voice starts with a sentence (I couldn’t catch exactly what it says), followed by "content =" before it begins the actual message. Could this be removed or adjusted? Also, I think the current speech pace is a bit fast—would it be possible to slow it down slightly for better clarity?
> 
> I agree with the decision not to add tests for this feature since it primarily complements the main functionality and doesn’t directly contribute to the UI or backend. Given the nature of the feature, adding tests would indeed be challenging and might not add significant value. Great call!
> 
> Great work overall! Maybe in the future, a good addition could be allowing the user to choose the voice they want to interact with. What do you think about this idea?
> 
> Looking forward to hearing your thoughts!

Hello Charbel,
Thanks for the remark about "content =", I fixed this in my latest commit, thanks for pointing it out! Personally I thought the voice was a bit fast at first but after trying a long interview it gets boring very quickly and so I feel like the speed is perfect on the long term. 
About adding different voices: I watched a few tutorials online and none explained how to change the voice: I think it would be a great task to add though! For now I think merging this PR to main and then making a new PR later where the user can select the voice of the interviewer would be best.

**Comment by charbelraffoul** [comment]
Hi Rudolf,

Thank you so much for taking my suggestions into consideration! I really appreciate it—overall this is a fantastic addition to our code.

I also tried it out for some time, and I have to agree with you—the current speed feels just right in the long term. It keeps the interaction smooth without getting too slow or boring.

I also noticed that you added tests—great job! It’s always reassuring to see the extra effort in ensuring quality and robustness.

I’m planning to accept this PR as soon as the merge conflicts are resolved. Keep up the awesome work, and I’m excited to see this feature integrated!

File: app/src/main/java/com/github/se/orator/MainActivity.kt (hunk 1 above)
Comment (charbelraffoul) [reply app/src/main/java/com/github/se/orator/MainActivity.kt:106]: After considering your perspective, I fully agree with your approach.

**Comment by baptisteghostine** [comment]
Hello guys, I need to have the latest version of the app to continue my work. Therefore I solved the conflicts of this branch and made sure that everything is okay. It passes the CI so I will merge now. Thank you for the great PR!

**Comment by baptisteghostine** [comment]
Hello guys, 

I noticed that the PR still needed changes. Since I need to have the latest version of the app to not have too much conflicts, I decided to fix the bugs and the merge conflicts with main myself. 

Thank you Charbel for noticing the bug, you can try to see that everything is okay. 

For a little overview of my changes: 

- In the MainActivity.kt I modify the arguments that OratorApp take to include both the changes on the main and the changes introduced by Rudolf. 
- I also modified the MainActivityTest.kt to pass the tests. 
- Finally, I modified the bug in ChatScreen.kt by disabling the AI voice when clicking on the back button, as mentioned. 

For the next PRs it would be great if some changes are requested, to fix them ASAP so that others can have the latest version of the code and continue working, but I understand that you were maybe busy during the weekend. 

Overall, very great PR! I would suggest for future improvements to add a feature to speak with the AI without clicking on record button to have a real conversation. It would be a great add-on!

**Comment by charbelraffoul** [comment]
Hi Baptiste,

Thanks for taking the time to address the pending changes and resolving the conflicts in the PR—it’s greatly appreciated! I’ve checked the updated app, and everything seems to be functioning smoothly.

I appreciate that you stepped in to handle the necessary updates, even though this wasn’t initially your PR. Your contributions, such as modifying the arguments in MainActivity.kt and addressing the bug in ChatScreen.kt with the AI voice, are solid improvements that enhance the user experience.

For future updates, the suggestion to add a feature for real-time conversations without needing to press the record button sounds intriguing and could take the app’s interactivity to the next level. Perhaps we can discuss potential implementations for this in the next sprint?

On another note, I was thinking that we could also consider using text-to-speech in the feedback screen in the future. It might make the feedback experience more engaging and consistent with the app’s interactive features.

Everything looks great, and I’m going to go ahead and accept this PR. Thanks again for your efforts on this PR, and I’m looking forward to collaborating on the next updates!

**Review by charbelraffoul — APPROVED** [review]

**Pull request merged**